---
title: "Format files for Eigensoft"
author: "Laura Tibbs"
date: "November 15, 2018"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## Goal: 
Edit genotype data to match that required for Eigensoft 3.0.

Set working directory to the location of this `.Rmd` file:
```{r}
# This needs to be changed to reflect the location of the GitHub repository and this code within it on any given computer that runs the code. 
# In my case, this is:
setwd("C:/Users/ltibbs/Box Sync/ISU/2018 fall/BCB 546X Computational Skills for Bio Data/BCB546X_MadMaize_final/code/PCA")
```

Load libraries:
```{r}
library(tidyverse)
library(readxl)
```

Import original genotype data from paper:
```{r}
full.geno <- read_csv("../../data/final_data.csv")
``` 

## Prepare data for Fig. 1: PCA on all accessions

### Make genotype file formatted for Eigensoft
According to the README on Eigensoft GitHub, this format is:
  "The genotype file contains 1 line per SNP.
  Each line contains 1 character per individual:
  0 means zero copies of reference allele.
  1 means one copy of reference allele.
  2 means two copies of reference allele.
  9 means missing data."
  
Inspect the data to see current format:

```{r}
head(full.geno)
```

Currently, each individual is in its own row, with its genotype at each of the SNPs recorded in the columns of that row as a pair of letters (e.g., "GG").

First, then, the tibble needs to be transposed so that each row contains one SNP rather than one individual.
```{r}
tr.geno <- as_tibble(cbind(SNPs=colnames(full.geno), t(full.geno))) # transpose full.geno and cbind the SNP names to it
colnames(tr.geno) <- tr.geno[1,] # add the column names since they were moved to row 1 by t()
tr.geno <- tr.geno[-1,] # remove row 1 now that column names are where they should be
head(tr.geno) # visually inspect transposed geno file
```

Next, I need to find the reference allele for each SNP.
```{r}
num.geno <- tr.geno %>%
  unite(united, -1, sep="", remove=F) %>%
  mutate(united_noNA=str_replace_all(united, "NA", ""), # remove NA values
         count.A=str_count(united_noNA, "A"), # count each nucleotide in the SNPs
         count.C=str_count(united_noNA,"C"),
         count.G=str_count(united_noNA, "G"),
         count.T=str_count(united_noNA,"T"),
         max.count=pmax(count.A, count.C, count.G, count.T), # find the max count of any NT for a given SNP
         ref.allele=ifelse(count.A==max.count, "A", # set the reference allele
                           ifelse(count.C==max.count, "C",
                                  ifelse(count.G==max.count, "G", "T")))
  ) %>%
  group_by(ref.allele) # group by reference allele value
```

And re-code the genotype data as the numeric 0,1,2, or 9.
```{r}
# split the genotype data into tibbles each with the same reference allele, and store these in a list
split.num.geno <- split(num.geno, num.geno$ref.allele)

# loop through each tibble in the list to replace homozygous reference as 2, het as 1, missing as 9, and other as 0:
for (i in 1:length(split.num.geno)) {
  ref <- unique(split.num.geno[[i]]$ref.allele) # save ref allele
  stopifnot(length(ref)==1) # check that there was only one ref allele in the table
  geno.index <- ncol(split.num.geno[[i]])-6 # store the index of the last column with genotype data
  
  # make string of heterozygous genotypes to match (code idea from https://stackoverflow.com/questions/35366803/how-to-use-gsub-for-the-list-of-multiple-patterns-in-r)
  het.list <- list(paste(ref, "A", sep=""), paste(ref, "C", sep=""), paste(ref, "G", sep=""), paste(ref, "T", sep=""),
                   paste("A", ref, sep=""), paste("C", ref, sep=""), paste("G", ref, sep=""), paste("T", ref, sep="")) # list possible het genotypes (refref genotype will also occur in this list, but will not be matched because that pattern will be replaced first)
  het.string <- paste(unlist(het.list), collapse="|") # convert this list into a string separated by | (logical or) to match any of these patterns
  
  split.num.geno[[i]] <- split.num.geno[[i]] %>% # in the following lines, use c(2:geno.index) to make replacements only in the coluns containing genotype data
  mutate_at(c(2:geno.index), funs(replace(., is.na(.), "9"))) %>% # replace missing data with 9
    mutate_at(c(2:geno.index), funs(str_replace_all(., paste(ref, ref, sep=""), "2"))) %>% # replace homozygous ref genotypes with 2
    mutate_at(c(2:geno.index), funs(str_replace_all(., het.string, "1"))) %>% # replace het genotypes with 1
    mutate_at(c(2:geno.index), funs(str_replace_all(., "[[:upper:]]{2}", "0"))) %>% # replace all remaining genotypes with 0
    select(-count.A, -count.C, -count.T, -count.G, -united, -united_noNA, -max.count) # remove non-genotype columns
}
```

Now, combine the split tibbles from the list and output genotype file for Eigensoft.
```{r}
eigensoft.geno <- do.call("rbind", split.num.geno) %>% # combine tibbles in list
  ungroup() %>%
  select(-ref.allele, -Sample) # remove unneeded columns

# save the full dataset in eigensoft format
full.eigensoft.geno <- unite(eigensoft.geno, united, sep="", remove=T) # collapse all SNPs into one line

# output file
write_tsv(full.eigensoft.geno, "full.eigensoft.geno", col_names = F)
```

### Make individual file formatted for Eigensoft
According to the README on Eigensoft GitHub, this format is:
  "The indiv file contains 1 line per individual.  There are 3 columns:
  1st column is sample ID.  Length is limited to 39 characters, including
    the family name if that will be concatenated.
  2nd column is gender (M or F).  If unknown, ok to set to U for Unknown.
  3rd column is a label which might refer to Case or Control status, or
    might be a population group label.  If this entry is set to "Ignore", 
    then that individual and all genotype data from that individual will be
    removed from the data set in all convertf output."
    
Read in the individuals' data:
```{r}
full.indiv <- read_tsv("../../data/12864_2015_1444_MOESM1_ESM.txt") # this file was attached as a ".csv" in the paper's supplemental data, but inspecting the file showed that it was actually a tsv file, so I manually changed the file extension
head(full.indiv)
```

Construct a file with the accession or line name (the column names of `eigensoft.geno`) in the first column, and U for unknown gender in the second column. Building this from the `eigensoft.geno` file ensures that these files are in the same order:
```{r}
pre.indiv <- tibble(line=colnames(eigensoft.geno), gender=c(rep("U", length(colnames(eigensoft.geno)))))
```

In Figure 1 of Blanca et al., the accessions are identified as one of SP, SC, SG, SLC, SLL, or mixture in the PC plot. These classifications are found in the `species` column of `full.indv`. This information can be put in the third column of the `.indiv` file, so join this information to the table:
```{r}
indiv.join <- left_join(pre.indiv, full.indiv, by=c("line"="Sample")) # join by sample ID
plyr::count(is.na(indiv.join$species)) # but not everything joined; why?

# 57 of the individual samples from full.indiv are NOT expected to join with our current dataset because they were removed as duplicates in data processing (see script in `code/Processing_Data`). This accounts for the 1008 samples in full.indiv vs 951 in pre.indiv (1008-951=57)

# After manual inspection, some of the other problems seem to be: 
# different cases (upper vs lower; lowercase in pre.indiv),
# periods vs commas (periods in pre.indiv), 
# and spaces vs underscores (underscores in pre.indiv)

# So, manually edit full.indiv so that it will match pre.indiv:
full.indiv$Sample <- gsub("\\,", "\\.", full.indiv$Sample) # switch commas to periods

# try to join now:
indiv.join <- left_join(pre.indiv, full.indiv, by=c("line"="Sample")) # join by sample ID
plyr::count(is.na(indiv.join$species)) # but not everything joined; why?

#Now only 3 are not joined, so manually edit these to match:
full.indiv$Sample <- gsub("PAS16398", "pas16398", full.indiv$Sample) # change case
full.indiv$Sample <- gsub("PAS16401", "pas16401", full.indiv$Sample) # change case
full.indiv$Sample <- gsub("PI 379051", "PI_379051", full.indiv$Sample) # change space to underscore

# try to join now:
indiv.join <- left_join(pre.indiv, full.indiv, by=c("line"="Sample")) # join by sample ID
plyr::count(is.na(indiv.join$species)) # Now everything has joined
stopifnot(sum(is.na(indiv.join$species))==0) # sanity check--stop if not everything has joined

```

Now, trim the file so that only the three needed columns are present, and output for Eigensoft.
```{r}
# extract needed columns
full.eigensoft.indiv <- indiv.join %>%
  select(line, gender, species) 

# output file
write_tsv(full.eigensoft.indiv, "full.eigensoft.indiv", col_names = F)
```

### Make snp file formatted for Eigensoft
According to the README on Eigensoft GitHub, this format is:
"The snp file contains 1 line per SNP.  There are 6 columns (last 2 optional):
  1st column is SNP name
  2nd column is chromosome.  X chromosome is encoded as 23.
    Also, Y is encoded as 24, mtDNA is encoded as 90, and XY is encoded as 91.
    Note: SNPs with illegal chromosome values, such as 0, will be removed
  3rd column is genetic position (in Morgans).  If unknown, ok to set to 0.0.
  4th column is physical position (in bases)
  Optional 5th and 6th columns are reference and variant alleles.
    For monomorphic SNPs, the variant allele can be encoded as X (unknown)."

Construct a file with the SNP name (the first columns of the tibbles in the list `split.num.geno`) in the first column and the second column (which will be moved to the third) all set to 0.0 because this column is optional. Building this from the `split.num.geno` list ensures that all files are in the same order.

```{r}
# Get SNP names
snp.names <- do.call("rbind", split.num.geno) %>% # combine tibbles in list
  ungroup() %>%
  select(SNP=Sample) # extract SNP names only, and rename

# add a column of zeroes:
pre.snp <- snp.names %>%
  mutate(cM=0.0)
```

Now, add chromosome and physical position for SNPs. According to the Blanca et al. paper, the maps from Sim et al. 2012 (https://doi.org/10.1371/journal.pone.0040563) were used. The physical positions and chromosome assignment of each SNP is found in the first supplemental table from that paper, which we have downloaded and placed in the `data` folder as `S1_SNPs_Sim_etal.xlsx`. Use this file to find the positions of SNPs in our dataset.

```{r}
# read in map file and tidy:
sim.snps <- read_xlsx("../../data/S1_SNPs_Sim_etal.xlsx")
# based on manual inspection of original file, column names have been moved to the first row. Fix this.
colnames(sim.snps) <- sim.snps[1,]
sim.snps <- sim.snps[-1,]
# select and rename needed columns
sim.snps <- sim.snps %>%
  select(SNP=`SNP Name`, Alt.SNP=`SolCAP SNP ID`, chr=Chromosome1, pos=`SNP Position (Mbp)1`)

# join with our snp file--looks like the Alt.SNP column is the one we need
full.snp <- left_join(pre.snp, sim.snps, by=c("SNP"="Alt.SNP"))
stopifnot(sum(is.na(full.snp$chr))==0) # check that it fully joined

# inspect and tidy the joined file:
plyr::count(full.snp$chr)
# there are SNPs with unknown chromosomes and positions; to format for Eigensoft, set all these chromosomes to 0
full.snp$pos[full.snp$chr=="Unknown"] <- 0 # if chr is unknown, set pos = 0
full.snp$chr[full.snp$chr=="Unknown"] <- 0 # if chr is unknown, set chr = 0 
# make columns numeric:
full.snp$chr <- as.numeric(full.snp$chr) 
full.snp$pos <- as.numeric(full.snp$pos) 
# keep only needed columns
full.snp <- select(full.snp, SNP, chr, cM, pos)

# sanity check: all cells have data?
stopifnot(sum(is.na(full.snp))==0)

```

Finally, format and output the data for Eigensoft. The file must have the columns in order as SNP name, chromosome, map position in Morgans, and physical position in bases. However, the physical position read in from the Sims et al. paper was in Mbp, not in bases. Therefore, I need to convert this column. One Mbp (mega base pair) is equal to 1,000,000 base pairs. I will also need to remove the column headers and save as a tab-delimited `.snp` file.

```{r}
# format data
full.snp <- full.snp %>%
  mutate(pos.bases=1000000*pos) %>% # convert position to bases
  select(SNP, chr, cM, pos.bases) # select needed columns

# write file for Eigensoft
write_tsv(full.snp, "eigensoft.snp", col_names = F)

```


## Prepare data for Fig. 2: PCA on SP, SLC, SLL, and SLL Mesoamerica + SLC subgroups

### Make individual files:

Fig. 2A contains SP lines (identified in `indiv.join$species`), colored by their subgroup membership (as shown in Table 2, Blanca et al 2015). Subgroups are found in `indiv.join$group2`. Similarly, Fig 2B contains SLC accessions and Fig 2C contains SLL accessions. and Fig 2D contains both SLC and SLL accessions, again colored by subgoup. According to the labels on Fig 2D, the groups in this figure are SLC Ecuador 3, SLC vintage, SLC Colombia, SLC Costa Rica, SLC Mesoamerica, SLL Mesoamerica, SLC world, SLC Peru 1, and SLC Peru 2. Looking at the data, these groups are found in `indiv.join$group2`, though with underscores substituted for spaces. 

So, I need to extract the individuals used to make each part of Figure 2 from the `indiv.join` dataset created above and use these to make new `.indiv` files.

```{r}
#Fig 2a: SP
SP.indiv <- indiv.join %>% 
  filter(species=="SP") %>% # filter SP lines
  select(line, gender, group2) # keep group2 to use to color the points in the figure later
write_tsv(SP.indiv, "SP.indiv", col_names=FALSE) # write file

# Fig 2b: SLC
SLC.indiv <- indiv.join %>% 
  filter(species=="SLC") %>% # filter SLC lines
  select(line, gender, group2) # keep group2 for later
write_tsv(SLC.indiv, "SLC.indiv", col_names=FALSE) # write file

# Fig 2c: SLL
SLL.indiv <- indiv.join %>% 
  filter(species=="SLL") %>% # filter SLL lines
  select(line, gender, group2) # keep group2 for later
write_tsv(SLL.indiv, "SLL.indiv", col_names=FALSE) # write file

# Fig 2d: SLL Mesoamerica and SLC groups
Fig2d.indiv <- indiv.join %>% 
  filter(group2 %in% c("SLC_Ecuador_3", "SLC_vintage", "SLC_Colombia", "SLC_Costa_Rica", 
                       "SLC_Mesoamerica", "SLL_Mesoamerica", "SLC_world", "SLC_Peru_1", "SLC_Peru_2")) %>% # filter lines
  select(line, gender, group2) # keep group2 for later
write_tsv(Fig2d.indiv, "Fig2d.indiv", col_names=FALSE) # write file

```

### Make genotype files:

Now, need to make genotype files that contain the same individuals as these four individual files do.

Do this by filtering out the individuals from the `eigensoft.geno` file created earlier.

```{r}
# Fig 2a: SP
SP.geno <- eigensoft.geno %>%
  select(SP.indiv$line)%>% # select those accessions in the indiv file
  unite(united, sep="", remove=T) # collapse all sNPs into one line
write_tsv(SP.geno, "SP.geno", col_names=F) # output file

# Fig 2b: SLC
SLC.geno <- eigensoft.geno %>%
  select(SLC.indiv$line) %>% # select those accessions in the indiv file
  unite(united, sep="", remove=T) # collapse all sNPs into one line
write_tsv(SLC.geno, "SLC.geno", col_names=F) # output file

# Fig 2c: SLL
SLL.geno <- eigensoft.geno %>%
  select(SLL.indiv$line)%>% # select those accessions in the indiv file
  unite(united, sep="", remove=T) # collapse all sNPs into one line
write_tsv(SLL.geno, "SLL.geno", col_names=F) # output file

# Fig 2d: SLL Mesoamerica and SLC groups
Fig2d.geno <- eigensoft.geno %>%
  select(Fig2d.indiv$line)%>% # select those accessions in the indiv file
  unite(united, sep="", remove=T) # collapse all sNPs into one line
write_tsv(Fig2d.geno, "Fig2d.geno", col_names=F) # output file
```

### SNP files:

The same SNP file will be used for all Eigensoft runs, because all groups use the same SNPs. Therefore, there is no need to make separate files for each part of Figure 2.
